### PR TITLE
Push down DELETE for enforceable filters on Delta Lake

### LINKED ISSUE #18331: Push down DELETE to the Delta Lake connector when filter is not applied on the data columns
https://trinodb.slack.com/archives/C03E2DYAS0J/p1689618089889359

```
trino> create table delta.tiny.ptest1(data integer, part integer) with (partitioned_by = array['part']);
CREATE TABLE
trino> insert into delta.tiny.ptest1 values (1,10), (2,10), (3,10), (11,20);
INSERT: 4 rows
trino> insert into delta.tiny.ptest1 values (5,10), (12,20);
INSERT: 2 rows

trino> delete from delta.tiny.ptest1 where part = 10;
```

When the delete filter is applied on a partition column, there is no need to do any data scanning on the Delta Lake table.
This should be a metadata only operation.

### Technical how-to
Implement `applyDelete`/`executeDelete` methods in `DeltaLakeMetadata`  (use Iceberg connector as proof of concept) to:

- Scan transaction log
- Flag files in the partition
- Append RemoveFile entries to the transaction log for each flagged file

In case of tables having CDF (Change Data Feed) enabled, the pushdown of the `DELETE` statement should be avoided.

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Perform DELETE only on the metadata of the Delta Lake
table when the delete filter does not touch any data
columns.

Sample queries affected by this enhancement are:

```
DELETE FROM table 
DELETE FROM table WHERE true
DELETE FROM partitioned_table WHERE part1=value1 AND part2=value2
DELETE FROM partitioned_table WHERE part2=value2
```

Fixes #18331

The operation will not be performed on the metadata layer when
there are `add` file entries in the transaction log of the table
without statistics containing number of records.

## Additional context and related issues

Without `DELETE` pushdown

```
trino:default> create table delta.tiny.sforders100 as select * from tpch.sf100.orders;
CREATE TABLE: 150000000 rows

trino:default> delete from delta.tiny.sforders100 where true;
DELETE: 150000000 rows

Query 20230725_125258_00003_xfv9h, FINISHED, 1 node
http://localhost:8080/ui/query.html?20230725_125258_00003_xfv9h
Splits: 25 total, 25 done (100.00%)
CPU Time: 115.1s total,  1.3M rows/s, 11.2MB/s, 89% active
Per Node: 1.1 parallelism, 1.46M rows/s, 12.5MB/s
Parallelism: 1.1
Peak Memory: 87.3MB
1:43 [150M rows, 1.26GB] [1.46M rows/s, 12.5MB/s]
```

With `DELETE` pushdown

```
trino:default> delete from delta.tiny.sf100orders where true;
DELETE: 150000000 rows

Query 20230725_123039_00000_fk4kt, FINISHED, 1 node
http://localhost:8080/ui/query.html?20230725_123039_00000_fk4kt
Splits: 1 total, 1 done (100.00%)
CPU Time: 0.1s total,     0 rows/s,     0B/s, 83% active
Per Node: 0.1 parallelism,     0 rows/s,     0B/s
Parallelism: 0.1
Peak Memory: 112B
1.84 [0 rows, 0B] [0 rows/s, 0B/s]
```

tldr; before 103 s after 1.8 s

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Improve performance of `DELETE` statement when it deletes the whole table or the filters apply only to the partition columns. ({issue}`18332 `)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/AbstractDeltaLakePageSink.java [hunk 1]
```diff
@@ -361,7 +361,7 @@ private int[] getWriterIndexes(Page page)
                 partitionName = Optional.of(partName);
             }
 
-            String fileName = session.getQueryId() + "-" + randomUUID();
+            String fileName = session.getQueryId() + "_" + randomUUID();
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/AbstractDeltaLakePageSink.java:364]: This change is in sync with https://github.com/trinodb/trino/blob/20e57792d27c2d98d758332978f16f0dc94baa28/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java#L333

I'd benefit from a suggestion in which utility class could i extract the method for creating the file name.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 2]
```diff
@@ -3116,6 +3117,86 @@ private Optional<SystemTable> getRawSystemTable(ConnectorSession session, Schema
...
 
+    @Override
+    public Optional<ConnectorTableHandle> applyDelete(ConnectorSession session, ConnectorTableHandle handle)
+    {
+        DeltaLakeTableHandle tableHandle = (DeltaLakeTableHandle) handle;
+        if (tableHandle.getMetadataEntry().isChangeDataFeedEnabled().orElse(false)) {
+            // For tables with CDF enabled the DELETE operation can't be performed only on metadata files
+            return Optional.empty();
+        }
+        if (!tableHandle.getNonPartitionConstraint().isAll()) {
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3128]: This check is probably not necessary - `applyDelete` is not being called when the table handle has a non partition constraint.

**Comment by findinpath** [comment]
`TestDeltaLakeDelete.testDeleteAllOssDeltaLake` is failing because the number of deleted rows is not anymore returned by the `DELETE` query.

Should the accelerated DELETE (done only based on table metadata) be performed if and only if all the data files have the number of records present in the transaction log file?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeExecuteDeleteTableHandle.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+import io.trino.spi.connector.ConnectorTableHandle;
+import io.trino.spi.connector.SchemaTableName;
+import io.trino.spi.predicate.TupleDomain;
+
+import java.util.List;
+import java.util.Objects;
+
+import static java.util.Objects.requireNonNull;
+
+public class DeltaLakeExecuteDeleteTableHandle
```
Comment (homar) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeExecuteDeleteTableHandle.java:29]: make it a record?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 4]
```diff
@@ -3116,6 +3117,95 @@ private Optional<SystemTable> getRawSystemTable(ConnectorSession session, Schema
...
+    }
+
+    private List<AddFileEntry> getAddFileEntriesMatchingEnforcedPartitionConstraint(ConnectorSession session, DeltaLakeTableHandle tableHandle)
+    {
+        TableSnapshot tableSnapshot;
+        try {
+            tableSnapshot = transactionLogAccess.loadSnapshot(tableHandle.getSchemaTableName(), tableHandle.getLocation(), session);
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (homar) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3196]: Shouldn't this be a TrinoException ?

**Review by homar — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 5]
```diff
@@ -3163,6 +3164,87 @@ private Optional<SystemTable> getRawSystemTable(SchemaTableName systemTableName)
...
+            TransactionLogWriter transactionLogWriter = transactionLogWriterFactory.newWriter(session, tableLocation);
+
+            long createdTime = Instant.now().toEpochMilli();
+            TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+            long currentVersion = getMandatoryCurrentVersion(fileSystem, tableLocation);
+            if (currentVersion != tableHandle.getReadVersion()) {
+                throw new TransactionConflictException(format("Conflicting concurrent writes found. Expected transaction log version: %s, actual version: %s", tableHandle.getReadVersion(), currentVersion));
+            }
+            long commitVersion = currentVersion + 1;
+            transactionLogWriter.appendCommitInfoEntry(getCommitInfoEntry(session, commitVersion, createdTime, MERGE_OPERATION, tableHandle.getReadVersion()));
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3199]: delete operation?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 6]
```diff
@@ -3163,6 +3164,87 @@ private Optional<SystemTable> getRawSystemTable(SchemaTableName systemTableName)
...
+
+            long writeTimestamp = Instant.now().toEpochMilli();
+
+            long deletedRecords = 0L;
+            boolean allDeletedFilesStatsPresent = true;
+            for (AddFileEntry addFileEntry : activeFiles) {
+                transactionLogWriter.appendRemoveFileEntry(new RemoveFileEntry(addFileEntry.getPath(), writeTimestamp, false));
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3206]: Data change feels like it should be true here

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 7]
```diff
@@ -3163,6 +3164,87 @@ private Optional<SystemTable> getRawSystemTable(SchemaTableName systemTableName)
...
+                Optional<Long> fileRecords = addFileEntry.getStats().flatMap(DeltaLakeFileStatistics::getNumRecords);
+                if (allDeletedFilesStatsPresent) {
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3208]: This felt more readable to me
```
allDeletedFileStatsPresent &= fileRecords.isPresent();
deletedRecords += fileRecords.orElse(0);
```

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java [hunk 8]
```diff
@@ -1780,6 +1780,55 @@ public void testTableWithTrailingSlashLocation(boolean partitioned)
...
+
+        assertUpdate(format("INSERT INTO %s (customer, purchases, address) VALUES ('Aaron', 5, 'Antioch'), ('Bill', 7, 'Antioch'), ('Mary', 10, 'Adelphi'), ('Aaron', 3, 'Dallas')", table), 4);
+
+        assertUpdate(format("DELETE FROM %s WHERE %s", table, deleteFilter), 2);
+        assertQuery("SELECT customer, purchases, address FROM " + table, "VALUES ('Mary', 10, 'Adelphi'), ('Aaron', 3, 'Dallas')");
+        assertUpdate("DROP TABLE " + table);
+    }
+
+    @DataProvider
+    public Object[][] deleteFiltersForTable()
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java:1797]: Please add one with a function call like String starts with...

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java [hunk 9]
```diff
@@ -223,7 +315,7 @@ public void testTableChangesFileSystemAccess()
         assertUpdate("INSERT INTO table_changes_file_system_access VALUES('url3', 'domain3', 3)", 1);
         assertUpdate("UPDATE table_changes_file_system_access SET page_url = 'url22' WHERE key = 'domain2'", 1);
         assertUpdate("UPDATE table_changes_file_system_access SET page_url = 'url33' WHERE views = 3", 1);
-        assertUpdate("DELETE FROM table_changes_file_system_access WHERE page_url = 'url1'", 1);
+        assertUpdate("DELETE FROM table_changes_file_system_access WHERE key = 'domain1'", 1);
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java:318]: This an intentional change?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java [hunk 10]
```diff
@@ -36,6 +38,34 @@
...
+        try {
+            onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES (1, 10), (2, 10), (11, 20), (21, 30), (22, 30)");
+            onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES (3, 10), (12, 20)");
+            if (partitioned) {
+                verify(onTrino().executeQuery("DELETE FROM default." + tableName + " WHERE p = 10").getRowsCount() == 3);
+                verify(onTrino().executeQuery("DELETE FROM default." + tableName).getRowsCount() == 4);
+            }
+            else {
+                verify(onTrino().executeQuery("DELETE FROM default." + tableName).getRowsCount() == 7);
+            }
```
Comment (alexjo2144) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java:62]: Verify the final state of the table onDelta

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 5 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3199]: I'm just following the same pattern as the one used for `UPDATE`/`DELETE`/`MERGE` statements in Delta Lake. See https://github.com/trinodb/trino/issues/15763

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java [hunk 11]
```diff
@@ -155,6 +158,95 @@ public void testReadWholePartition()
...
+                        .addCopies(new FileOperation(TRANSACTION_LOG_JSON, "00000000000000000003.json", INPUT_FILE_EXISTS), 1)
+                        .addCopies(new FileOperation(TRANSACTION_LOG_JSON, "00000000000000000003.json", INPUT_FILE_NEW_STREAM), 4) // TODO (https://github.com/trinodb/trino/issues/16780) why is last transaction log accessed more times than others?
+                        .addCopies(new FileOperation(TRANSACTION_LOG_JSON, "00000000000000000004.json", INPUT_FILE_NEW_STREAM), 1)
+                        .build());
+
+        assertUpdate("DROP TABLE test_delete_part_key");
+    }
+
+    @Test
+    public void testDeleteWholeTable()
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java:191]: Can we merge this `testDeleteWholeTable` with `testDeleteWholePartition` and parametrize query? Maybe add `DELETE FROM table` case?

**Comment by findinpath** [comment]
Rebasing on `master` to address conflicts

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java (hunk 11 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java:191]: For `testDeleteWholeTable` i've changed the test to use a non-partitioned table. I guess it is more revealing this way.

**Comment by findinpath** [comment]
@pajaks  & @alexjo2144  AC. Please review again the PR.

**Review by alexjo2144 — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/AbstractDeltaLakePageSink.java (hunk 1 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/AbstractDeltaLakePageSink.java:360]: You can always add a new one ;)

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 2 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3270]: if this is not necessary then remove it :P

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java [hunk 12]
```diff
@@ -1820,6 +1820,65 @@ public void testTableWithTrailingSlashLocation(boolean partitioned)
         assertUpdate("DROP TABLE " + tableName);
     }
 
+    @Test(dataProvider = "deleteFiltersForTable")
+    public void testDeleteWithFilter(String createTableSql, String deleteFilter)
```
Comment (homar) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java:1824]: Didn't we have tests that test deletion previously? If we did what is the point of these? They don't test the feature only correctness of deletion

**Review by homar — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 13]
```diff
@@ -3258,6 +3259,80 @@ public WriterScalingOptions getInsertWriterScalingOptions(ConnectorSession sessi
...
+            return Optional.empty();
+        }
+
+        return Optional.of(tableHandle);
+    }
+
+    @Override
+    public OptionalLong executeDelete(ConnectorSession session, ConnectorTableHandle handle)
+    {
+        DeltaLakeTableHandle tableHandle = (DeltaLakeTableHandle) handle;
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3280]: Please throw an exception when `isAppendOnly` returns true. `TestDeltaLakeDeleteCompatibility` should be updated to cover this secneario.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 14]
```diff
@@ -3258,6 +3259,80 @@ public WriterScalingOptions getInsertWriterScalingOptions(ConnectorSession sessi
...
+            TransactionLogWriter transactionLogWriter = transactionLogWriterFactory.newWriter(session, tableLocation);
+
+            long writeTimestamp = Instant.now().toEpochMilli();
+            TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+            long currentVersion = getMandatoryCurrentVersion(fileSystem, tableLocation);
+            if (currentVersion != tableHandle.getReadVersion()) {
+                throw new TransactionConflictException(format("Conflicting concurrent writes found. Expected transaction log version: %s, actual version: %s", tableHandle.getReadVersion(), currentVersion));
+            }
+            long commitVersion = currentVersion + 1;
+            transactionLogWriter.appendCommitInfoEntry(getCommitInfoEntry(session, commitVersion, writeTimestamp, MERGE_OPERATION, tableHandle.getReadVersion()));
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3294]: We can use `DELETE_OPERATION` instead of `MERGE_OPERATION` in my opinion. 

Please udpate a comment of `DELETE_OPERATION` constant if you change.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 15]
```diff
@@ -3258,6 +3259,80 @@ public WriterScalingOptions getInsertWriterScalingOptions(ConnectorSession sessi
...
+            for (AddFileEntry addFileEntry : activeFiles) {
+                transactionLogWriter.appendRemoveFileEntry(new RemoveFileEntry(addFileEntry.getPath(), writeTimestamp, true));
+
+                Optional<Long> fileRecords = addFileEntry.getStats().flatMap(DeltaLakeFileStatistics::getNumRecords);
+                allDeletedFilesStatsPresent &= fileRecords.isPresent();
+                deletedRecords += fileRecords.orElse(0L);
+            }
+
+            transactionLogWriter.flush();
+            Optional<Long> checkpointInterval = tableHandle.getMetadataEntry().getCheckpointInterval();
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3307]: nit: Inline `checkpointInterval` variable.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 16]
```diff
@@ -3258,6 +3259,80 @@ public WriterScalingOptions getInsertWriterScalingOptions(ConnectorSession sessi
...
+
+    private List<AddFileEntry> getAddFileEntriesMatchingEnforcedPartitionConstraint(ConnectorSession session, DeltaLakeTableHandle tableHandle)
+    {
+        TableSnapshot tableSnapshot;
+        try {
+            tableSnapshot = transactionLogAccess.loadSnapshot(tableHandle.getSchemaTableName(), tableHandle.getLocation(), session);
+        }
+        catch (IOException | RuntimeException e) {
+            throw new TrinoException(DELTA_LAKE_INVALID_SCHEMA, "Error getting snapshot for " + tableHandle.getSchemaTableName(), e);
+        }
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3324]: Why not call the existing `getSnapshot` method?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java (hunk 12 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java:1824]: We may want to assert the plan like below:
```java
        assertUpdate(
                getSession(),
                format("DELETE FROM %s WHERE %s", table, deleteFilter),
                2,
                plan -> {
                    boolean tableDelete = searchFrom(plan.getRoot()).where(node -> node instanceof TableDeleteNode).matches();
                    assertTrue(tableDelete, "...");}
        );
```

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java [hunk 17]
```diff
@@ -1820,6 +1820,65 @@ public void testTableWithTrailingSlashLocation(boolean partitioned)
...
+                {
+                        // delete filter applied only partly on first partitioned field
+                        "CREATE TABLE %s (purchases INT, address VARCHAR, customer VARCHAR) WITH (location = 's3://%s/%s', partitioned_by = ARRAY['address', 'customer'])",
+                        "address = 'Antioch'"
+                },
+                {
+                        // delete filter applied only partly on second partitioned field
+                        "CREATE TABLE %s (purchases INT, address VARCHAR, customer VARCHAR) WITH (location = 's3://%s/%s', partitioned_by = ARRAY['customer', 'address'])",
+                        "address = 'Antioch'"
+                }
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java:1878]: Extra `,` will reduce the diff when adding the new condition in the future:
```suggestion
                },
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java [hunk 18]
```diff
@@ -33,9 +34,42 @@
...
+    @Flaky(issue = DATABRICKS_COMMUNICATION_FAILURE_ISSUE, match = DATABRICKS_COMMUNICATION_FAILURE_MATCH)
+    public void testDeleteOnEnforcedConstraintsReturnsRowsCount(boolean partitioned)
+    {
+        String tableName = "test_delete_push_down_" + randomNameSuffix();
+        onDelta().executeQuery("" +
+                "CREATE TABLE default." + tableName +
+                "         (v INT, p INT)" +
+                "         USING delta " +
+                (partitioned ? "         PARTITIONED BY (p)" : "") +
+                "         LOCATION 's3://" + bucketName + "/databricks-compatibility-test-" + tableName + "'");
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java:50]: nti: I would remove redundant spaces:
```java
        onDelta().executeQuery("" +
                "CREATE TABLE default." + tableName +
                "(v INT, p INT)" +
                "USING delta " +
                (partitioned ? "PARTITIONED BY (p)" : "") +
                "LOCATION 's3://" + bucketName + "/databricks-compatibility-test-" + tableName + "'");
```

**Comment by ebyhr** [comment]
Pushed small changes to fix CI failure.
